File: src/main/java/me/sd_master92/customvoting/listeners/PlayerListener.java
package me.sd_master92.customvoting.listeners;

import me.sd_master92.customvoting.Main;
import me.sd_master92.customvoting.VoteFile;
import me.sd_master92.customvoting.constants.Data;
import me.sd_master92.customvoting.constants.Settings;
import me.sd_master92.customvoting.constants.enumerations.SoundType;
import me.sd_master92.customvoting.database.PlayerRow;
import me.sd_master92.customvoting.gui.VotePartyRewards;
import me.sd_master92.customvoting.subjects.CustomVote;
import me.sd_master92.customvoting.subjects.VoteParty;
import me.sd_master92.customvoting.subjects.VoteTopSign;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Sign;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;

public class PlayerListener implements Listener
{
    public static List<UUID> moneyInput = new ArrayList<>();
    public static List<UUID> commandInput = new ArrayList<>();
    private final Main plugin;

    public PlayerListener(Main plugin)
    {
        this.plugin = plugin;
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event)
    {
        Player player = event.getPlayer();
        if (plugin.useDatabase())
        {
            List<String> queue = new ArrayList<>();
            PlayerRow playerRow = new PlayerRow(plugin, player);
            for (int i = 0; i < playerRow.getQueue(); i++)
            {
                queue.add("unknown.com");
            }
            executeQueue(queue, player);
            playerRow.clearQueue();
        } else
        {
            VoteFile voteFile = new VoteFile(player, plugin);
            executeQueue(voteFile.getQueue(), player);
            voteFile.clearQueue();
        }

        if (player.isOp() && plugin.getConfig().getBoolean(Settings.INGAME_UPDATES) && !plugin.isUpToDate())
        {
            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    plugin.sendDownloadUrl(player);
                    player.sendMessage("");
                    player.sendMessage(ChatColor.GRAY + "Updates can be turned off in the /votesettings");
                }
            }.runTaskLater(plugin, 20 * 5);
        }
    }

    private void executeQueue(List<String> queue, Player player)
    {
        if (!queue.isEmpty())
        {
            plugin.print(queue.size() + " queued votes found for " + player.getName() + ". Forwarding in 10 seconds.." +
                    ".");
            Iterator<String> iterator = queue.iterator();
            new BukkitRunnable()
            {
                @Override
                public void run()
                {
                    if (iterator.hasNext())
                    {
                        CustomVote.create(plugin, player.getName(), iterator.next());
                    } else
                    {
                        cancel();
                    }
                }
            }.runTaskTimer(plugin, 200L, 20L);
        }
    }

    @EventHandler
    public void onCommandProcess(PlayerCommandPreprocessEvent event)
    {
        Player player = event.getPlayer();
        if (moneyInput.contains(player.getUniqueId()))
        {
            event.setCancelled(true);
            player.sendMessage(ChatColor.RED + "Enter a number");
        }
        if (commandInput.contains(player.getUniqueId()))
        {
            event.setCancelled(true);
            String command = event.getMessage();
            checkCommand(command, player);
        }
    }

    @EventHandler
    public void onPlayerChat(AsyncPlayerChatEvent event)
    {
        Player player = event.getPlayer();
        if (moneyInput.contains(player.getUniqueId()))
        {
            event.setCancelled(true);
            if (event.getMessage().equalsIgnoreCase("cancel"))
            {
                moneyInput.remove(player.getUniqueId());
                SoundType.FAILURE.play(plugin, player);
            } else
            {
                try
                {
                    double input = Double.parseDouble(event.getMessage());
                    if (input < 0 || input > 1000000)
                    {
                        player.sendMessage(ChatColor.RED + "Enter a number between 0 and 1.000.000");
                    } else
                    {
                        plugin.getConfig().set(Settings.VOTE_REWARD_MONEY, input);
                        plugin.getConfig().saveConfig();
                        moneyInput.remove(player.getUniqueId());
                        SoundType.SUCCESS.play(plugin, player);
                        player.sendMessage(ChatColor.GREEN + "Successfully updated the money reward!");
                    }
                } catch (Exception e)
                {
                    player.sendMessage(ChatColor.RED + "Enter a number");
                }
            }
        }
        if (commandInput.contains(player.getUniqueId()))
        {
            event.setCancelled(true);
            if (event.getMessage().equalsIgnoreCase("cancel"))
            {
                commandInput.remove(player.getUniqueId());
                SoundType.FAILURE.play(plugin, player);
            } else
            {
                checkCommand(event.getMessage(), player);
            }
        }
    }

    @EventHandler
    public void onBlockBreak(BlockBreakEvent event)
    {
        Player player = event.getPlayer();
        Block block = event.getBlock();
        if (block.getType() == Material.ENDER_CHEST)
        {
            Map<String, Location> chests = plugin.getData().getLocations(Data.VOTE_PARTY);
            for (String key : chests.keySet())
            {
                if (chests.get(key).equals(event.getBlock().getLocation()))
                {
                    if (player.hasPermission("çustomvoting.voteparty"))
                    {
                        if (plugin.getData().deleteLocation(Data.VOTE_PARTY + "." + key))
                        {
                            plugin.getData().deleteItems(Data.VOTE_PARTY + "." + key);
                            event.setDropItems(false);
                            event.getPlayer().sendMessage(ChatColor.RED + "Vote Party Chest #" + key + " deleted.");
                        }
                    } else
                    {
                        event.setCancelled(true);
                        player.sendMessage(ChatColor.RED + "You do not have permission to break this block.");
                    }
                }
            }
        } else if (block.getState() instanceof Sign)
        {
            VoteTopSign voteTop = VoteTopSign.get(block.getLocation());
            if (voteTop != null)
            {
                voteTop.delete(player);
            }
        } else
        {
            List<Location> locations = new ArrayList<>();
            for (BlockFace face : new BlockFace[]{BlockFace.UP, BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH,
                    BlockFace.WEST})
            {
                if (block.getRelative(face).getState() instanceof Sign)
                {
                    locations.add(block.getRelative(face).getLocation());
                }
            }
            if (!locations.isEmpty())
            {
                for (Location loc : locations)
                {
                    VoteTopSign voteTop = VoteTopSign.get(loc);
                    if (voteTop != null)
                    {
                        voteTop.delete(player);
                    }
                }
            }
        }
    }

    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event)
    {
        if (event.getItemInHand().equals(VoteParty.VOTE_PARTY_ITEM))
        {
            Player player = event.getPlayer();
            if (player.hasPermission("customvoting.voteparty"))
            {
                Set<String> chests = plugin.getData().getLocations(Data.VOTE_PARTY).keySet();
                int i = 1;
                while (chests.contains("" + i))
                {
                    i++;
                }
                plugin.getData().setLocation(Data.VOTE_PARTY + "." + i, event.getBlock().getLocation());
                SoundType.SUCCESS.play(plugin, player);
                player.sendMessage(ChatColor.GREEN + "Vote Party Chest #" + i + " registered.");
                player.getInventory().setItemInMainHand(VoteParty.VOTE_PARTY_ITEM);
            } else
            {
                event.setCancelled(true);
                player.sendMessage(ChatColor.RED + "You do not have permission to place this block.");
            }
        }
    }

    @EventHandler
    public void onBlockInteract(PlayerInteractEvent event)
    {
        Player player = event.getPlayer();
        if (event.getAction() == Action.RIGHT_CLICK_BLOCK && event.getClickedBlock() != null && event.getClickedBlock().getType() == Material.ENDER_CHEST && !player.isSneaking())
        {
            Location loc = event.getClickedBlock().getLocation();
            Map<String, Location> chests = plugin.getData().getLocations(Data.VOTE_PARTY);
            for (String key : chests.keySet())
            {
                if (chests.get(key).equals(loc))
                {
                    event.setCancelled(true);
                    if (player.hasPermission("customvoting.voteparty"))
                    {
                        SoundType.OPEN.play(plugin, player);
                        player.openInventory(new VotePartyRewards(plugin, key).getInventory());
                    } else
                    {
                        player.sendMessage(ChatColor.RED + "You do not have permission to open this chest.");
                    }
                }
            }
        }
    }

    private void checkCommand(String command, Player player)
    {
        for (String forbidden : plugin.getConfig().getStringList(Settings.FORBIDDEN_COMMANDS))
        {
            if (command.toLowerCase().contains(forbidden))
            {
                player.sendMessage(ChatColor.RED + "This command is forbidden.");
                return;
            }
        }
        if (command.startsWith("/"))
        {
            command = command.substring(1);
        }
        List<String> commands = plugin.getData().getStringList(Data.VOTE_COMMANDS);
        if (commands.contains(command))
        {
            commands.remove(command);
            player.sendMessage(ChatColor.RED + "Removed /" + command + " from commands");
        } else
        {
            commands.add(command);
            player.sendMessage(ChatColor.GREEN + "Added /" + command + " to commands");
        }
        plugin.getData().set(Data.VOTE_COMMANDS, commands);
        plugin.getData().saveConfig();
        commandInput.remove(player.getUniqueId());
    }
}

File: src/main/kotlin/me/sd_master92/customvoting/listeners/PlayerListener.kt
package me.sd_master92.customvoting.listeners

import com.github.shynixn.mccoroutine.bukkit.launch
import me.sd_master92.core.errorLog
import me.sd_master92.core.infoLog
import me.sd_master92.core.inventory.ConfirmGUI
import me.sd_master92.core.tasks.TaskTimer
import me.sd_master92.customvoting.CV
import me.sd_master92.customvoting.constants.enumerations.*
import me.sd_master92.customvoting.constants.interfaces.Voter
import me.sd_master92.customvoting.gui.items.VotePartyItem
import me.sd_master92.customvoting.gui.pages.editors.VotePartyRewardItemsEditor
import me.sd_master92.customvoting.gui.pages.menus.VotePartyCrate
import me.sd_master92.customvoting.helpers.ParticleHelper
import me.sd_master92.customvoting.sendText
import me.sd_master92.customvoting.stripColor
import me.sd_master92.customvoting.subjects.CustomVote
import me.sd_master92.customvoting.subjects.VoteCrate
import me.sd_master92.customvoting.subjects.VoteTopSign
import me.sd_master92.customvoting.subjects.stands.VoteTopStand
import me.sd_master92.customvoting.subjects.voteparty.VoteParty
import me.sd_master92.customvoting.subjects.voteparty.VotePartyChest
import me.sd_master92.customvoting.tasks.UpdateChecker
import me.sd_master92.customvoting.tasks.VoteReminder
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.BlockFace
import org.bukkit.block.Sign
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent

class PlayerListener(private val plugin: CV) : Listener
{
    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent)
    {
        val player = event.player
        VoteReminder.remindPlayer(plugin, player)
        UpdateChecker.checkUpdates(plugin, player)
        plugin.launch {
            executeQueue(player)
        }
    }

    @EventHandler
    @Suppress("UNUSED_PARAMETER")
    fun onPlayerLeave(event: PlayerQuitEvent)
    {
        TaskTimer.delay(plugin, 20)
        {
            VoteTopStand.updateAll(plugin)
        }.run()
    }

    @EventHandler
    fun onWorldChange(event: PlayerChangedWorldEvent)
    {
        plugin.launch {
            executeQueue(event.player)
        }
    }

    private suspend fun executeQueue(player: Player)
    {
        val voter = Voter.get(plugin, player)
        val queue = voter.getHistory().filter { it.queued }
        if (!voter.clearQueue())
        {
            plugin.errorLog(PMessage.QUEUE_ERROR_DELETE_XY.with(player.uniqueId.toString(), player.name))
        }
        if (queue.isNotEmpty())
        {
            plugin.infoLog(PMessage.QUEUE_MESSAGE_FORWARD_XY.with(queue.size.toString(), player.name))
            val iterator = queue.iterator()
            TaskTimer.repeat(plugin, 20, 200)
            {
                if (iterator.hasNext())
                {
                    CustomVote.create(plugin, player.name, iterator.next().site.toString(), true)
                } else
                {
                    it.cancel()
                }
            }.run()
        }
    }

    @EventHandler
    fun onPlayerInteractEntity(event: EntityDamageByEntityEvent)
    {
        val entity = event.entity
        val damager = event.damager

        if (entity is ArmorStand && damager is Player)
        {
            for (voteTopStand in VoteTopStand.voteTops.values)
            {
                if (voteTopStand.topStand.stand?.uniqueId == entity.uniqueId ||
                    voteTopStand.nameStand.stand?.uniqueId == entity.uniqueId ||
                    voteTopStand.votesStand.stand?.uniqueId == entity.uniqueId
                )
                {
                    event.isCancelled = true
                    if (!plugin.config.getBoolean(Setting.DISABLED_MESSAGE_ARMOR_STAND.path))
                    {
                        damager.sendText(plugin, Message.VOTE_TOP_STANDS_DONT)
                    }
                }
            }
        }
    }

    @EventHandler
    fun onBlockBreak(event: BlockBreakEvent)
    {
        val player = event.player
        val block = event.block
        if (block.type == Material.ENDER_CHEST)
        {
            val votePartyChest = VotePartyChest.getByLocation(plugin, event.block.location)
            if (votePartyChest != null)
            {
                if (player.hasPermission("çustomvoting.voteparty"))
                {
                    event.isDropItems = false
                    votePartyChest.deleteLocation(player)
                } else
                {
                    event.isCancelled = true
                    player.sendMessage(PMessage.ACTION_ERROR_BREAK_BLOCK_NO_PERMISSION.toString())
                }
            }
            val voteCrate = VoteCrate.getByLocation(plugin, event.block.location)
            if (voteCrate != null)
            {
                if (player.hasPermission("customvoting.crate"))
                {
                    voteCrate.deleteLocation(player)
                } else
                {
                    event.isCancelled = true
                    player.sendMessage(PMessage.ACTION_ERROR_BREAK_BLOCK_NO_PERMISSION.toString())
                }
            }
        } else if (block.state is Sign)
        {
            val voteTop = VoteTopSign[block.location]
            voteTop?.delete(player)
        } else
        {
            val locations: MutableList<Location> = ArrayList()
            for (face in arrayOf(
                BlockFace.UP, BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH,
                BlockFace.WEST
            ))
            {
                if (block.getRelative(face).state is Sign)
                {
                    locations.add(block.getRelative(face).location)
                }
            }
            if (locations.isNotEmpty())
            {
                for (loc in locations)
                {
                    val voteTop = VoteTopSign[loc]
                    voteTop?.delete(player)
                }
            }
        }
    }

    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent)
    {
        val item = event.itemInHand
        if(event.blockPlaced.type == Material.ENDER_CHEST)
        {
            val key = item.itemMeta?.displayName?.split("#")?.reversed()?.get(0)
            if (key != null && item.itemMeta!!.displayName == (VotePartyItem(key).itemMeta?.displayName))
            {
                val player = event.player
                if (player.hasPermission("customvoting.voteparty"))
                {
                    val chest = VotePartyChest.getByKey(plugin, key)
                    if(chest != null)
                    {
                        if(chest.loc == null)
                        {
                            chest.loc = event.block.location
                            player.sendMessage(PMessage.VOTE_PARTY_MESSAGE_CHEST_LOCATION_SET_X.with(key))
                        } else
                        {
                            event.isCancelled = true
                            SoundType.FAILURE.play(plugin, player)
                            player.sendMessage(PMessage.VOTE_PARTY_MESSAGE_CHEST_LOCATION_SET_ALREADY_X.with(key))
                        }
                    } else
                    {
                        VotePartyChest.create(plugin, player, key)?.loc = event.block.location
                    }
                } else
                {
                    event.isCancelled = true
                    player.sendMessage(PMessage.ACTION_ERROR_BREAK_BLOCK_NO_PERMISSION.toString())
                }
            }
        }
    }

    @EventHandler
    fun onBlockInteract(event: PlayerInteractEvent)
    {
        val player = event.player
        val clicked = event.clickedBlock

        if (event.action != Action.RIGHT_CLICK_BLOCK || clicked == null)
        {
            return
        }

        if (clicked.type == Material.ENDER_CHEST)
        {
            val votePartyChest = VotePartyChest.getByLocation(plugin, clicked.location)
            if (votePartyChest != null)
            {
                event.isCancelled = true

                if (VoteParty.IS_ACTIVE && VoteParty.getCurrent()?.votePartyType == VotePartyType.LOCKED_CRATES && !votePartyChest.isOpened)
                {
                    votePartyChest.isOpened = true
                    votePartyChest.hide(votePartyChest.loc!!)
                    SoundType.OPEN.play(plugin, player)
                    VotePartyCrate(plugin, votePartyChest).open(player)

                    plugin.broadcastText(Message.VOTE_PARTY_CHEST_CLAIMED, mapOf(Pair("%PLAYER%", player.name)))
                    ParticleHelper.shootFirework(plugin, player.location)

                    if (VotePartyChest.getAll(plugin).none { chest -> !chest.isOpened })
                    {
                        VoteParty.stop(plugin)
                    }
                    return
                } else
                {
                    if (player.hasPermission("customvoting.voteparty"))
                    {
                        SoundType.OPEN.play(plugin, player)
                        VotePartyRewardItemsEditor(plugin, null, votePartyChest.key).open(player)
                    } else
                    {
                        player.sendMessage(PMessage.ACTION_ERROR_OPEN_CHEST_NO_PERMISSION.toString())
                    }
                    return
                }
            }
        }

        var voteCrate = VoteCrate.getByLocation(plugin, clicked.location)
        if (voteCrate != null && (event.item?.type != Material.TRIPWIRE_HOOK || event.item?.itemMeta?.lore?.get(0)
                ?.contains("#") == false)
        )
        {
            event.isCancelled = true
            player.sendMessage(PMessage.ACTION_ERROR_OPEN_CHEST_NEED_KEY.toString())
            return
        }

        if (!event.hasItem() || event.item!!.type != Material.TRIPWIRE_HOOK)
        {
            return
        }

        val item = event.item!!
        val lore = item.itemMeta?.lore?.get(0)

        if (lore == null || !lore.contains("#"))
        {
            return
        }

        val key = lore.split("#")[1].stripColor()
        voteCrate = VoteCrate.getByKey(plugin, key)

        if (voteCrate == null)
        {
            return
        }

        val meta = item.itemMeta!!
        meta.setDisplayName(PMessage.CRATE_ITEM_NAME_KEY_X.with(voteCrate.name))
        item.itemMeta = meta
        player.inventory.setItemInMainHand(item)

        if (voteCrate.loc == null)
        {
            if (player.hasPermission("customvoting.crate"))
            {
                event.isCancelled = true
                val loc = clicked.location
                val otherVoteCrate = VoteCrate.getByLocation(plugin, loc)

                if (otherVoteCrate == null)
                {
                    val confirm = object : ConfirmGUI(
                        plugin,
                        PMessage.CRATE_INVENTORY_NAME_PLACE_CONFIRM_X.with(voteCrate.name)
                    )
                    {
                        override fun onCancel(event: InventoryClickEvent, player: Player)
                        {
                            SoundType.FAILURE.play(plugin, player)
                            player.closeInventory()
                        }

                        override fun onClose(event: InventoryCloseEvent, player: Player)
                        {
                            SoundType.FAILURE.play(plugin, player)
                        }

                        override fun onConfirm(event: InventoryClickEvent, player: Player)
                        {
                            voteCrate.create(loc, player)
                        }
                    }

                    SoundType.CLICK.play(plugin, player)
                    confirm.open(player)
                } else
                {
                    SoundType.FAILURE.play(plugin, player)
                    player.sendMessage(PMessage.ACTION_ERROR_OPEN_CHEST_WRONG_KEY.toString())
                }
            } else
            {
                player.sendMessage(PMessage.ACTION_ERROR_OPEN_CHEST_NO_PERMISSION.toString())
            }
        } else if (clicked.location == voteCrate.loc)
        {
            event.isCancelled = true
            item.amount--
            player.inventory.setItemInMainHand(item)
            voteCrate.open(player)
        } else
        {
            event.isCancelled = true
            player.sendMessage(PMessage.ACTION_ERROR_INTERACT_NEED_CRATE.toString())
        }
    }
}